File: as/src/test/java/com/trikset/gamepad/SenderServiceTest.java
package com.trikset.gamepad;

import static org.junit.Assert.assertEquals;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;
import org.robolectric.annotation.LooperMode;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.ServerSocket;
import java.net.Socket;

@RunWith(RobolectricTestRunner.class)
@LooperMode(LooperMode.Mode.LEGACY)
@Config(sdk = {Config.OLDEST_SDK, Config.TARGET_SDK, Config.NEWEST_SDK})
public class SenderServiceTest {
    @Test
    public void senderServiceShouldConnectToServerSuccessfullyAfterSendingCommand()
            throws InterruptedException {
        DummyServer server = new DummyServer(1, 4);
        SenderService client = new SenderService();
        client.setTarget(DummyServer.IP, DummyServer.DEFAULT_PORT + 4);

        client.send("");
        Robolectric.flushBackgroundThreadScheduler();
        Assert.assertTrue(server.isConnected());
    }

    @Test
    public void senderServiceShouldSendSingleCommandCorrectly() throws InterruptedException {
        DummyServer server = new DummyServer(1, 0);
        SenderService client = new SenderService();
        client.setTarget(DummyServer.IP, DummyServer.DEFAULT_PORT);
        client.setKeepaliveTimeout(10000000); // to disable keep-alive messages

        client.send("Test; check");
        Robolectric.flushBackgroundThreadScheduler();
        assertEquals("Test; check", server.getLastCommand());
    }

    @Test
    public void senderServiceShouldSendMultipleCommandsCorrectly() throws InterruptedException {
        DummyServer server = new DummyServer(5, 1);
        SenderService client = new SenderService();
        client.setTarget(DummyServer.IP, DummyServer.DEFAULT_PORT + 1);
        client.setKeepaliveTimeout(10000000); // to disable keep-alive messages

        for (int i = 0; i < 5; ++i) {
            client.send(String.format("%d checking", i));
        }
        Robolectric.flushBackgroundThreadScheduler();
        assertEquals("4 checking", server.getLastCommand());
    }

    @Test
    public void setTargetShouldSetServerSuccessfully() {
        SenderService client = new SenderService();
        client.setTarget("someaddr-test", 12345);

        assertEquals("someaddr-test", client.getHostAddr());
    }

    @Test
    public void senderServiceShouldReturnCorrectKeepaliveTimeout() {
        SenderService client = new SenderService();

        client.setKeepaliveTimeout(3453);
        assertEquals(3453, client.getKeepaliveTimeout());

        client.setKeepaliveTimeout(1234);
        assertEquals(1234, client.getKeepaliveTimeout());
    }

    private static class DummyServer {
        static final String IP = "localhost";
        static final int DEFAULT_PORT = 12345;

        final int port;

        private boolean isConnected = false;
        public boolean isConnected() { return isConnected; }

        private String lastCommand;
        public String getLastCommand() { return lastCommand; }


        DummyServer(final int cmdNumber, final int portShift) {
            port = DEFAULT_PORT + portShift;

            Thread serverThread = new Thread(() -> {
                try (ServerSocket server = new ServerSocket(port)) {
                    Socket client = server.accept();
                    isConnected = true;

                    BufferedReader clientInput =
                            new BufferedReader(new InputStreamReader(client.getInputStream()));
                    for (int i = 0; i < cmdNumber; ++i) {
                        lastCommand = clientInput.readLine();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            });
            serverThread.start();
        }
    }
}

File: app/src/test/kotlin/com/trikset/gamepad2/SenderServiceTest.kt
package com.trikset.gamepad2

import android.os.Looper.getMainLooper
import java.util.Locale
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows.shadowOf
import org.robolectric.android.util.concurrent.PausedExecutorService
import org.robolectric.annotation.LooperMode
import org.robolectric.annotation.LooperMode.Mode.PAUSED

@RunWith(RobolectricTestRunner::class)
@LooperMode(PAUSED)
class SenderServiceTest : RobolectricTestBase() {
  private val mExecutor = PausedExecutorService()
  private var client: SenderService? = null

  // A connected client keeps a real keepalive scheduler thread alive. Disconnect
  // every client in @After so no keepalive task can land on a later test.
  @After
  fun tearDown() {
    client?.disconnect("tearDown")
  }

  @Test
  fun senderServiceShouldConnectToServerSuccessfullyAfterSendingCommand() {
    TestTcpServer().use { server ->
      client = SenderService(mExecutor)
      client!!.setTarget(TestTcpServer.HOST, server.port)
      client!!.send("")
      mExecutor.runAll()
      shadowOf(getMainLooper()).idle()
      assertTrue(server.awaitConnection())
      assertTrue(server.isConnected)
    }
  }

  @Test
  fun senderServiceShouldSendSingleCommandCorrectly() {
    TestTcpServer().use { server ->
      client = SenderService(mExecutor)
      client!!.setTarget(TestTcpServer.HOST, server.port)
      client!!.keepaliveTimeout = 10000000 // to disable keep-alive messages
      client!!.send("Test; check")
      mExecutor.runAll()
      shadowOf(getMainLooper()).idle()
      assertTrue(server.awaitCount(1))
      assertEquals("Test; check", server.lastMessage())
    }
  }

  @Test
  fun senderServiceShouldSendMultipleCommandsCorrectly() {
    TestTcpServer().use { server ->
      client = SenderService(mExecutor)
      client!!.setTarget(TestTcpServer.HOST, server.port)
      client!!.keepaliveTimeout = 10000000 // to disable keep-alive messages

      for (i in 0 until 5) {
        client!!.send(String.format(Locale.ROOT, "%d checking", i))
      }
      mExecutor.runAll()
      shadowOf(getMainLooper()).idle()
      assertTrue(server.awaitCount(5))
      assertEquals("4 checking", server.lastMessage())
    }
  }

  @Test
  fun setTargetShouldSetServerSuccessfully() {
    client = SenderService(mExecutor)
    client!!.setTarget("someaddr-test", 0)
    assertEquals("someaddr-test", client!!.hostAddr)
  }

  @Test
  fun connectShouldEstablishConnectionWithoutSending() {
    TestTcpServer().use { server -> establishConnection(server) }
  }

  @Test
  fun connectWhenAlreadyConnectedShouldNotReconnect() {
    TestTcpServer().use { server ->
      establishConnection(server)
      // A second connect while connected (mOut != null) must not drop/re-open.
      client!!.connect()
      mExecutor.runAll()
      assertTrue(server.isConnected)
    }
  }

  /** Connects [client] to [server] via the public connect() entry point and asserts the state. */
  private fun establishConnection(server: TestTcpServer): SenderService {
    client = SenderService(mExecutor)
    client!!.setTarget(TestTcpServer.HOST, server.port)
    client!!.connect()
    mExecutor.runAll()
    shadowOf(getMainLooper()).idle()
    assertTrue(server.awaitConnection())
    assertTrue(server.isConnected)
    return client!!
  }

  @Test
  fun senderServiceShouldReturnCorrectKeepaliveTimeout() {
    client = SenderService(mExecutor)
    client!!.keepaliveTimeout = 3453
    assertEquals(3453, client!!.keepaliveTimeout)
    client!!.keepaliveTimeout = 1234
    assertEquals(1234, client!!.keepaliveTimeout)
  }

  @Test
  fun connectWithBlankOrNullHostShouldNotConnect() {
    client = SenderService(mExecutor)
    // Fresh sender: host is null -> the isNullOrBlank guard blocks the connect.
    client!!.connect()
    mExecutor.runAll()
    assertTrue(client!!.connectionState.value is ConnectionState.Disconnected)
    // Blank host (video-only) is equally a no-op.
    client!!.setTarget("", 4444)
    client!!.connect()
    mExecutor.runAll()
    assertTrue(client!!.connectionState.value is ConnectionState.Disconnected)
  }

  @Test
  fun sendAndDisconnectShouldCompleteTheSendPath() {
    TestTcpServer().use { server ->
      client = SenderService(mExecutor)
      client!!.setTarget(TestTcpServer.HOST, server.port)
      client!!.keepaliveTimeout = 10000000
      client!!.send("ping")
      mExecutor.runAll()
      shadowOf(getMainLooper()).idle()
      assertTrue(server.awaitCount(1))
      client!!.disconnect("done")
    }
  }
}
